authz/audit_logger.go

```diff
@@ -0,0 +1,90 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package authz
+
+import (
+	"sync"
+)
+
+var (
+	mu sync.Mutex
+	m  = make(map[string]AuditLoggerBuilder)
+)
+
+// RegisterAuditLoggerBuilder registers the builder in a global map
+// using b.Name() as the key.
+// This should only be called during initialization time (i.e. in an init() function).
+// If multiple builders are registered with the same name, the one registered last
+// will take effect.
+func RegisterAuditLoggerBuilder(b AuditLoggerBuilder) {
+	mu.Lock()
+	defer mu.Unlock()
+	m[b.Name()] = b
+}
+
+// AuditInfo contains information used by the audit logger during an audit logging event.
+type AuditInfo struct {
+	// RPCMethod is the method of the audited RPC.
+	RPCMethod string `json:"rpc_method,omitempty"`
+	// Principal is the identity of the RPC. Currently it will only be available in
+	// certificate-based TLS authentication.
+	Principal string `json:"principal,omitempty"`
+	// PolicyName is the authorization policy name (or the xDS RBAC filter name).
+	PolicyName string `json:"policy_name,omitempty"`
+	// MatchedRule is the matched rule (or policy name in the xDS RBAC filter). It will be
+	// empty if there is no match.
+	MatchedRule string `json:"matched_rule,omitempty"`
+	// Authorized indicates whether the audited RPC is authorized or not.
+	Authorized bool `json:"bool,omitempty"`
+}
+
+// AuditLoggerConfig defines the configuration for a particular implementation of audit logger.
+type AuditLoggerConfig interface {
+	// Name returns the same name as that returned by its supported builder.
+	Name() string
+}
+
+// AuditLogger is the interface for an audit logger.
+type AuditLogger interface {
+	// Log logs the auditing event with the given information.
+	Log(*AuditInfo)
+}
+
+// AuditLoggerBuilder is the interface for an audit logger builder.
+type AuditLoggerBuilder interface {
+	// ParseAuditLoggerConfig parses an implementation-specific config into a
+	// structured logger config this builder can use to build an audit logger.
+	ParseAuditLoggerConfig(config interface{}) (AuditLoggerConfig, error)
+	// Build builds an audit logger with the given logger config.
+	Build(AuditLoggerConfig) AuditLogger
```

Comment:
Can this not error?  What if the logger is supposed to write to a file but the directory doesn't exist, for example?

Reply:
If `Build()` returns an error at this stage, the logger will be nil but the data plane will have no way to NACK the filter. I think the invalidation you mentioned above should be done by `ParseAuditLoggerConfig` and I added comments to explain that.

This is consistent with other languages in dealing with things like LB policies. We will only call this `Build` with the config returned from `ParseAuditLoggerConfig` and implementers must ensure a valid config from there is a valid one for building a logger.

Reply:
I see what you're saying about nacking the config, but there are failure modes caused by runtime problems even with a valid config.  Is the implication that if there is an error like this, the logger needs to keep retrying to recover?  Or should it go into a lame mode and just do nothing forever?  Is there some guidance we want to provide for potential users here, and should it go into the gRFC/a user guide/docstrings (in all languages)?

Reply:
Per our offline discussion, I think we agreed that the builder should never fail to create a logger given a validated config. Runtime issues should be handled by the logger impl. I documented this in the gRFC.